backpack/core/derivatives/batchnorm_nd.py

```diff
@@ -0,0 +1,327 @@
+
+        return einsum("c,vnc->vnc", factor, r_mat)
+
+    ###############################################################
+    #                  HELPER FUNCTIONS                         ###
+    ###############################################################
+    def _get_normalized_input_and_var(
+        self, module: Union[BatchNorm1d, BatchNorm2d, BatchNorm3d]
+    ) -> Tuple[Tensor, Tensor]:
+        input: Tensor = module.input0
+        if module.training:
+            dim: Tuple[int] = self._get_free_axes(module, index_batch=0)
+            mean: Tensor = input.mean(dim=dim)
+            var: Tensor = input.var(dim=dim, unbiased=False)
+        else:
+            mean: Tensor = module.running_mean
+            var: Tensor = module.running_var
+        mean: Tensor = self._unsqueeze_free_axis(module, mean, 0)
+        var_expanded: Tensor = self._unsqueeze_free_axis(module, var, 0)
+        return (input - mean) / (var_expanded + module.eps).sqrt(), var
+
+    def _get_denominator(
+        self, module: Union[BatchNorm1d, BatchNorm2d, BatchNorm3d]
+    ) -> int:
+        shape_input: Size = module.input0.shape
+        free_axes: Tuple[int] = self._get_free_axes(module, index_batch=0)
+        denominator: int = 1
+        for index in free_axes:
+            denominator *= shape_input[index]
+        return denominator
+
+    @staticmethod
+    def _get_n_axis(module: Union[BatchNorm1d, BatchNorm2d, BatchNorm3d]) -> int:
+        return module.input0.dim() - 2
+
+    def _unsqueeze_free_axis(
+        self,
+        module: Union[BatchNorm1d, BatchNorm2d, BatchNorm3d],
+        tensor: Tensor,
+        index_batch: int,
+    ) -> Tensor:
+        """Unsqueezes the free dimensions.
+
+        This function is useful to avoid broadcasting.
+        Also useful when applying .expand(self._get_free_axes()) afterwards.
+
+        Args:
+            module: extended module
+            tensor: the tensor to operate on
+            index_batch: the batch axes index
+
+        Returns:
+            tensor with the free dimensions unsqueezed.
+        """
+        out = tensor.unsqueeze(index_batch)
+        for _ in range(self._get_n_axis(module)):
+            out = out.unsqueeze(-1)
+        return out
+
+    def _get_free_axes(
+        self,
+        module: Union[BatchNorm1d, BatchNorm2d, BatchNorm3d],
+        with_batch_axis: bool = True,
+        index_batch: int = 1,
+    ) -> Tuple[int]:
+        free_axes: List[int] = []
+        if with_batch_axis:
+            free_axes.append(index_batch)
+        for n in range(self._get_n_axis(module)):
+            free_axes.append(index_batch + n + 2)
+        return tuple(free_axes)
+
+    free_index: Dict[int, str] = {
+        0: "",
+        1: "l",
+        2: "hw",
+        3: "dhw",
+    }
+
+    free_index_alternative: Dict[int, str] = {
+        0: "",
+        1: "x",
+        2: "xy",
+        3: "xyz",
+    }
```

Comment:
If the above suggestions work, we should be able to get rid of one dict.

Reply:
Worked out. I removed both dictionaries since the last one was used only a single time.